### ISSUE TITLE
[kube-prometheus-stack] To insert secrets into Alertmanager config (not config as a secret) 

### ISSUE THREAD

**Issue by andrew-pickin-epi** [prometheus-community/helm-charts#412]
There is no easy way not getting the contents of secrets into Alertmanager configuration. If there is it is not documented well.

Some fields are listed as `<tmpl_secret>` which might help, but this too is poorly documented. How for example does one get the contents of a secret into such a field? But more specifically some obvious fields are listed a `<secret>` (which does not appear to have any functional difference only cosmetic, i.e. they are hidden) eg. smtp/email config authentication.
I suggest these would better referred to as hidden rather than overloading the term secret.

There appears some reticence to use environment variables as this is seen as insecure. However, this is the de facto standard way of achieving configuration from secrets it is commonplace throughout published helm charts. What is more, both Prometheus and Grafana support this mechanism. This approach makes prefect the the enemy of the good: it encourages people to include sensitive information within their source code. And when all is said and done one needs access to the pod to view the contents of environment variables.

The alternative is to make the whole (or part) of the configuration a secret. This is less that idea and complicates workflow, it make iteration with things like Flux and sealed secrets more problematic and again encourages secrets (or at least base64 encoded copies into source code again).  A config file needs to be generated every time the secret changes, so this now needs to be templates outside of helm, and the changes are that this is done by using an environment variable! 

So it might be possible to extend the `<secret>` functionality to interpolate a secret value. But is not very flexible. What is the field I'm trying to protect isn't tagged as such?

So, if I'm missing something and the solution already exists - it's not documented. 
I suggest supporting environment variables as [kubernetes suggests](https://kubernetes.io/docs/concepts/configuration/secret/#using-secrets-as-environment-variables) at least until a better solution is provided.

**Comment by tobypeschel** [comment]
I agree completely with this. I've been searching for a way to include `smtp_auth_password` in the alertmanager config without it being plaintext for a long time. There doesn't appear to be any way to achieve this, which is crazy. 

+1 for the idea of supporting environment variables.

**Comment by monotek** [comment]
Would on of you guys mind to create an pr for this?

**Comment by begemotik** [comment]
Would love to see someone getting it done.

**Comment by slapula** [comment]
alertmanager has a long history of avoiding this industry standard feature (see https://github.com/prometheus/alertmanager/issues/504 and https://github.com/prometheus/prometheus/issues/2357).  However,
there is a new experimental feature in Prometheus that allows environment variable expansion via external labels: https://www.lisenet.com/2021/use-external-labels-with-prometheus-alerts/.  I'm no Helm expert so I'm not quite sure how to make that work here (or if we want codify something experimental like this just yet) but I'll play around with it.

I too am supremely annoyed that I have commit my `smtp_auth_password` in plain text.  It makes me wonder how many people have accidentally committed it (or something similar) to public repos.

**Comment by antoineozenne** [comment]
Here is my workaround to avoid passing `smtp_auth_password` in the command line or storing the alertmanager configuration file entirely in a secret: use an `AlertmanagerConfig` CR. It allows using a `Secret` for the `smtp_auth_password`.

For example, use these values to deploy the `kube-prometheus-stack` chart:

```yaml
alertmanager:
  alertmanagerSpec:
    alertmanagerConfigNamespaceSelector: # Merge all AlertmanagerConfig CR in this namespace with the initial config of alertmanager
      matchLabels:
        name: kube-prometheus-stack
  config: # The minimal necessary configuration (initial config)
    global:
      resolve_timeout: 5m
```

Then manually apply the `alertmanagerconfig.yaml` file:

```yaml
apiVersion: monitoring.coreos.com/v1alpha1
kind: AlertmanagerConfig
metadata:
  name: kube-prometheus-stack-alertmanagerconfig
  namespace: kube-prometheus-stack
  labels:
    app.kubernetes.io/instance: kube-prometheus-stack
    app.kubernetes.io/name: alertmanagerconfig
spec:
  inhibitRules: []
  route:
    receiver: "null"
    routes:
      - groupBy:
          - email_to
        receiver: email_router
        groupWait: 0s
        groupInterval: 5m
        repeatInterval: 2h
  receivers:
    - name: email_router
      emailConfigs:
        - sendResolved: true
          smarthost: XXXXXXX
          requireTLS: true
          authUsername: XXXXXXX
          authPassword:
            name: secretName # The secret you can inject into the namespace any way you want
            key: secretKey
          to: XXXXXXX
          from: XXXXXXX
          html: XXXXXXX
```

This method makes it easy to deploy in a GitOps mode, with Flux, for example. To improve that, kube-prometheus-stack should allow `AlertmanagerConfig` CRs to be declared directly in chart values (some charts I've seen allow resources of any type to be declared directly in values, just in case...).

**Comment by tobypeschel** [comment]
@antoineozenne Only just saw your reply, thank you! I will definitely utilise this. 👍🏻

**Comment by andrew-pickin-epi** [comment]
An alternative method for configuring Alertmanager -> smtp gateway is to use a postfix pod. (eg eeacms/postfix).
This means no authentication is required inside the cluster to get mail off AM, and your smtp gw creds only need to be in one place: postfix. Postfix behaves well with secrets.
This means 

- your whole AM config doesn't need to be secret.
- you can see all mail leaving the cluster in one place (or one sub-system if you have replicas)

But it doesn't solve the problem for other stuff.

**Comment by colans** [comment]
Just do this with Terraform.  [Terraform can set its own variables from your existing environment](https://developer.hashicorp.com/terraform/cli/config/environment-variables#tf_var_name), and then use these for setting values in Helm charts.  e.g.

```terraform
resource "helm_release" "kubernetes_prometheus_stack" {
  name       = "kube-prometheus-stackr"
  namespace  = var.kubernetes_prometheus_namespace
  version    = var.kubernetes_prometheus_stack_version
  repository = "https://prometheus-community.github.io/helm-charts"
  chart      = "kube-prometheus-stack"
  depends_on = [kubernetes_namespace.monitoring]
  set {
    name  = "prometheus.prometheusSpec.retention"
    value = "${var.prometheus_data_keep_days}d"
  }
  [...]
  set {
    name  = "alertmanager.config.receivers[0].email_configs[0].auth_password"
    value = var.prometheus_alertmanager_smtp_password
  }
}

You should be using Terraform anyway, because it automates all of your IaC.

**Comment by timvahlbrock** [comment]
As far as I can tell there is still not a more practical solution for this. I had some problems getting @antoineozenne workaround to work, but eventually got it running (maybe because I switched from `alertmanagerConfigNamespaceSelector` to `alertmanagerConfigSelector`?). Another thing that I realized when checking why my rules weren't working: By default, when using an additional `AlertmanagerConfig` the alertmanager will add a matcher for the origin namespace to all the contained routes (so namespace owners can configure routes for their namespace with it, without allowing them to leak other namespaces alerts to their own notifications). One can change this behavior by setting `alertmanager.alertmanagerSpec.alertmanagerConfigMatcherStrategy.type` to `None` in the kube-prometheus-stack values.
